tests/BaGetter.Core.Tests/Metadata/RegistrationBuilderTests.cs

```diff
@@ -1,75 +1,110 @@
+            Generator.GetPackage(packageId, "3.1.0-pre"),
+            Generator.GetPackage(packageId, "1.0.0-beta1"),
+            Generator.GetPackage(packageId, "1.0.0"),
         };
 
-        /*
-         * Use current date for each packages publish date, because later a date offset will be
-         * calculated and leads to an overflow error of the offset because the default is year 0001.
-         */
-        foreach (var package in packages)
-        {
-            package.Published = DateTime.UtcNow;
-        }
+        return new PackageRegistration(packageId, packages);
+    }
+
+    #endregion
 
-        var registration = new PackageRegistration(packageId, packages);
+    [Fact]
+    public void Ctor_UrlGeneratorIsNull_ShouldThrow()
+    {
+        // Act/Assert
+        var ex = Assert.Throws<ArgumentNullException>(() => new RegistrationBuilder(null));
+    }
 
+    [Fact]
+    public void BuildIndex_PackageRegistrationIsNull_ShouldThrow()
+    {
+        // Arrange
         var registrationBuilder = new RegistrationBuilder(_urlGenerator.Object);
 
+        // Act/Assert
+        var ex = Assert.Throws<ArgumentNullException>(() => registrationBuilder.BuildIndex(null));
+    }
+
+    [Fact]
+    public void BuildIndex_RegistrationIndexResponse_ShouldBeSortedByVersion()
+    {
+        // Arrange
+        var registration = GetPackageRegistration();
+
         // Act
-        var response = registrationBuilder.BuildIndex(registration);
+        var response = _registrationBuilder.BuildIndex(registration);
 
         // Assert
-        Assert.Equal(packages.Count, response.Pages[0].ItemsOrNull.Count);
+        Assert.Equal(registration.Packages.Count, response.Pages[0].ItemsOrNull.Count);
+
         var index = 0;
-        foreach (var package in packages.OrderBy(p => p.Version))
+        foreach (var package in registration.Packages.OrderBy(p => p.Version))
         {
             Assert.Equal(package.Version.ToFullString(), response.Pages[0].ItemsOrNull[index++].PackageMetadata.Version);
         }
     }
 
-    /// <summary>
-    /// Create a fake <see cref="Package"></see> with the minimum metadata needed by the <see cref="RegistrationBuilder"></see>.
-    /// </summary>
-    private Package GetTestPackage(string packageId, string version)
+    [Fact]
+    public void BuildIndex_RegistrationIndexResponse_ShouldHaveCorrectTotalDownloads()
     {
-        return new Package
-        {
-            Id = packageId,
-            Authors = new string[] { "test" },
-            PackageTypes = new List<PackageType> { new PackageType { Name = "test" } },
-            Dependencies = new List<PackageDependency> { },
-            Version = new NuGetVersion(version),
-        };
+        // Arrange
+        var registration = GetPackageRegistration();
+
+        // Act
+        var response = _registrationBuilder.BuildIndex(registration);
+
+        // Assert
+        Assert.Equal(registration.Packages.Count, response.TotalDownloads);
```

Comment:
`registration.Packages.Count` would contain the number of different version of this package, wouldn't it? This does not conceptually equal the total downloads of the package. There could be 6 versions of the package, yet thousands of downloads.

Reply:
You are correct.
```
//RegistrationBuilder.cs
TotalDownloads = registration.Packages.Sum(p => p.Downloads),
```